lib/types/json-schema.ts

```diff
@@ -88,20 +88,24 @@ export type JSONSchemaType<T, _partial extends boolean = false> = (
           // "properties" are optional for more concise dictionary schemas
           // "patternProperties" and can be only used with interfaces that have string index
           type: JSONType<"object", _partial>
-          // "required" type does not guarantee that all required properties are listed
-          // it only asserts that optional cannot be listed
-          required: _partial extends true ? Readonly<(keyof T)[]> : Readonly<RequiredMembers<T>[]>
           additionalProperties?: boolean | JSONSchemaType<T[string]>
           unevaluatedProperties?: boolean | JSONSchemaType<T[string]>
           properties?: _partial extends true ? Partial<PropertiesSchema<T>> : PropertiesSchema<T>
           patternProperties?: {[Pattern in string]?: JSONSchemaType<T[string]>}
-          propertyNames?: JSONSchemaType<string>
+          propertyNames?: Omit<JSONSchemaType<string>, "type"> & {type?: "string"}
           dependencies?: {[K in keyof T]?: Readonly<(keyof T)[]> | PartialSchema<T>}
           dependentRequired?: {[K in keyof T]?: Readonly<(keyof T)[]>}
           dependentSchemas?: {[K in keyof T]?: PartialSchema<T>}
           minProperties?: number
           maxProperties?: number
-        }
+        } & (// "required" type does not guarantee that all required properties
+        // are listed it only asserts that optional cannot be listed.
+        // "required" is not necessary if it's a non-partial type with no required keys
+        _partial extends true
+          ? {required: Readonly<(keyof T)[]>}
```

Comment:
I am wondering - why "required" is required in partial schemas? I am probably missing something, as I write the example, it does not seem required (even when it has to be). This example type checks:

```typescript
type MyUnion = {tag: "a"; a: number} | {tag: "b"; b: number}

const myUnionSchema: JSONSchemaType<MyUnion> = {
  type: "object",
  required: ["tag"],
  oneOf: [
    {
      properties: {
        tag: {const: "a"},
        a: {type: "number"},
      },
    },
    {
      properties: {
        tag: {const: "b"},
        b: {type: "number"},
      },
    },
  ],
};
```

but it probably it should have `required: ["a"]` (or "b") in subschemas...

Either way, it's unrelated issue - I am going to merge it as is - thank you very much!

Reply:
I think it doesn't matter because because partial schemas are just that, partial:

https://github.com/ajv-validator/ajv/blob/c9e0676094e04011b64187b2ec3b747a86198cd7/lib/types/json-schema.ts#L4